### PR TITLE
refactor: update invoice registry to latest standards

### LINKED ISSUE #1749: refactor: Invoice registry
References issue https://github.com/IMA-WorldHealth/bhima-2.X/issues/1737

The patient registry should be updated to meet the latest standards in order to reduce technical debt and give clients required features: 

- [ ] Use the latest FilterService in favour of the hardcoded deprecated method
  - [ ] Update the InvoiceService to use FIilterService 
  - [ ] Update the search modal to use 'default' and 'custom' filters
- [ ] Integrate grid save state (https://github.com/IMA-WorldHealth/bhima-2.X/issues/1738)
  - [ ] Save configuration 
  - [ ] Load default configuration
- [ ] "Menu" dropdown for core utilities
  * Column configuration
  * Download view as CSV
  * Download view as PDF 
- [ ] Remove bhRenderer dropdown 
- [ ] Use bhClear on search modal (https://github.com/IMA-WorldHealth/bhima-2.X/issues/1702)
- [ ] Remove 'Print' action favouring the 'Download as PDF'
- [ ] Ensure links to other registries work correctly

### PR DESCRIPTION
**Pull request by jniles**
This PR updates the Invoice Registry to the latest standard.  It should help developers of other registries figure out how to update their registries as well.

~~**Note**: The 5K+ line change is because I've committed a package-lock.json file that comes by default with npm v5+.  You can find it in the [release notes](https://github.com/npm/npm/releases/tag/v5.0.0).  If you would prefer me to remove it and us to ignore it, that's cool.~~ I've removed the `package-lock.json` as it appears to cause issues with previous NPM versions.

For reference of what changed, all components of #1749 were tackled here.

#### Code Changes
 1. I've refined the `FilterParser` implementation on the server by turning off `autoParseStatements` (see #1454).  Every filter must be specified.  I've also changed cashReference comparison statement to convert the binary uuid in JS, since this is what most other modules do.
 2. I've migrated the filters to using the client `FilterService`.
 3. I've ensured the filter modal to latest standards.  This includes using `bhClear` (see #1702).
 4. The end-to-end tests have been updated to reflect these changes.  They now use a combined `SearchModal` service, shared with the Cash Payments Registry.  It provides simple control of the filters.
 5. Two options for downloading have been added: PDF and CSV.  For uniformity, they both use the server to serve files to an HTML `<a href download>` tag.  If needed, I can try to implement the CSV rendering with the Exporter service ... but I worry about the difference in UX.  I'll look to review for guidance.
 6. I've implemented the GridState service on this grid.  It works as the Journal's works.
 7. I've implemented the Grid Column service.  It also uses the GridState as described above... however, there is a bug in the Grid Column service reported in #1759.

#### A Note on End to End Tests
I've combined the Cash Payment Registry's `SearchModal` page object with the Invoice Registry's `SearchModal`.  Basically, any search form should be able to leverage this page object without writing their own massive test suite.  It prefers convention over configuration though, so you'll have to name your controllers `$ctrl`, your search parameters `searchQueries` (as done in the Journal and Cash Payments registry) and your default search paramters `defaultQueries`.

#### Screenshots
Pics or it didn't happen, right?

**New User Interface**
invoiceregistrynewui

**New Menu Demo**
invoiceregistrydropdownmenu

**Download as CSV Option**
invoiceregistrydownloadascsv png

**Column Configuration Modal**
invoiceregistrycolumnconfigurationmodal

**Column Changes Saved in Save State**
invoiceregistrycolumnconfigurationsaved

Closes #1749.

----
Thank you for contributing!

Before submitting this pull request, please verify that you have:
 - [x] Run your code through ESLint.  [Check out our styleguide](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/docs/STYLEGUIDE.md).
 - [x] Run integration tests.
 - [x] Run end-to-end tests.
 - [x] Accurately described the changes your are making in this pull request.

For a more detailed checklist, [see the official review checklist](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub) that this PR will be evaluated against.

Ensuring that the above checkboxes are completed will help speed the review process and help build a stronger application.  Thanks!

### CODE REVIEW

**Comment by jniles** [comment]
Before this gets merged, I need to figure out how to show the grid footer on the same page.  I think it's just a CSS class on the grid. :+1:

**Comment by sfount** [comment]
This looks like a great refactor! From the feature review this morning there was only one comment: 

* Many of our grids/ registries overflow on the bottom of the page so that you cannot see the footer row - we should add to all issues that this should be resolved on each registry.

**Comment by jniles** [comment]
@sfount, changes have been made.

File: client/src/modules/invoices/registry/registry.html [hunk 1]
```diff
@@ -6,66 +6,83 @@
     </ol>
 
     <div class="toolbar">
+
+      <div class="toolbar-item">
+        <div uib-dropdown dropdown-append-to-body data-action="open-menu">
+          <a class="btn btn-default" uib-dropdown-toggle>
+            <span class="fa fa-bars"></span> <span class="hidden-xs" translate>FORM.LABELS.MENU</span> <span class="caret"></span>
```
Comment (sfount) [inline client/src/modules/invoices/registry/registry.html:13]: This is a good standard 👍

**Review by sfount — APPROVED** [review]
Code looks good to me. Changes from this morning's functionality review have been factored in. 

This will be a good standard to reference from the upcoming work on registries.

**Pull request merged**